src/IntelliTect.Coalesce/Utilities/StringExtensions.cs

```diff
@@ -82,31 +58,157 @@ public static string ToProperCase(this string theString)
             return result;
         }
 
+        public static string EscapeStringLiteralForLinqDynamic(this string str) => str?
+            .Replace(@"\", @"\\")
+            .Replace("\"", "\"\"");
+
+        public static string EscapeStringLiteralForCSharp(this string str) => str?
+            .Replace(@"\", @"\\")
+            .Replace("\"", "\\\"");
+
         /// <summary>
-        /// Adds the items in the colleciton to list if they aren't already there.
+        /// Convert a string to a valid C# identifier that conforms to language specification 2.4.2
         /// </summary>
-        public static void AddUnique(this List<string> list, IEnumerable<string> items)
+        /// <param name="string">The string to convert</param>
+        /// <param name="prefix">(Optional) string to prefix the identifier with.</param>
+        /// <returns>A valid C# identifier</returns>
+        public static string GetValidCSharpIdentifier(this string @string, string prefix = null)
         {
-            foreach (var item in items)
+            if (!string.IsNullOrWhiteSpace(prefix))
             {
-                list.AddUnique(item);
+                @string = prefix + @string;
             }
+            @string = @string.Trim();
+
+            @string = Regex.Replace(@string, @"(?!\\u[0-9A-Fa-f]{4})[^\w]", match =>
+            {
+                if (match.Index == 0 && match.Value == "@")
+                    return "@";
+                return @"_";
+            });
+
+            if (!Regex.IsMatch(@string, "^[_@a-zA-Z]"))
+            {
+                @string = $"_{@string}";
+            }
+            else if (IsCSharpKeyword(@string))
+            {
+                @string = $"@{@string}";
+            }
+
+            return @string;
         }
 
-        /// <summary>
-        /// Adds the items in the colleciton to list if they aren't already there.
-        /// </summary>
-        public static void AddUnique(this List<string> list, string item)
+        private static bool IsCSharpKeyword(string @string)
```

Comment:
This could be replaced with https://github.com/dotnet/roslyn/blob/1b0cf5c732062f66b71a3d62a165d6eb5f8b3022/src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs#L1124 and https://github.com/dotnet/roslyn/blob/1b0cf5c732062f66b71a3d62a165d6eb5f8b3022/src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs#L788

Reply:
Done